frontend-web/src/coffee-call/call-aggregation.ts

```diff
@@ -0,0 +1,71 @@
+import { delay, map, merge, Observable, of, scan, switchMap, tap } from "rxjs";
+import { ReceivedCallMessage } from "./protocol";
+
+export const enum CoffeeCallState {
+  inactive,
+  announced,
+  inProgress,
+  canceled,
+};
+
+export class CoffeeCall {
+  state: CoffeeCallState = CoffeeCallState.inactive;
+  participants: Set<string> = new Set();
+
+  constructor(public messages: ReceivedCallMessage[]) {
+    if (messages.length === 0) {
+      return;
+    }
+
+    for (const message of messages) {
+      switch (message.type) {
+        case 'start':
+          this.state = CoffeeCallState.inProgress;
+          this.participants.add(message.name);
+          break;
+        case 'join':
+          if (this.state !== CoffeeCallState.inProgress) {
+            this.state = CoffeeCallState.announced;
+          }
+          this.participants.add(message.name);
+          break;
+        case 'leave':
+          this.participants.delete(message.name);
+          break;
+        default:
+          assertUnreachable(message.type);
+      }
+    }
```

Comment:
This _really_ screams "state machine"!

We might want to take a look at [XState](https://xstate.js.org/) (probably not right now).

Reply:
As we have discussed, this is a state machine of some sort (although we're only interested in its terminal state not in the transitions ;)). But would you really introduce a dependency for just these 15loc? Are you trying to outnumber @ohaz and with his 530 rust dependencies? :P

Reply:
Well, it's more declarative (compare to rxjs) and also this whole project is to learn about stuff, and I always wanted to play around with an FSM lib.

If size is a concern (it is not), [Robot is a more minimal alternative](https://thisrobot.life/guides/comparison-with-xstate.html#bundle-size) to XState:
> Minified and gzipped XState is 13kB at the time of this writing. Robot is 1kB.

Reply:
Yes, that is a good argument ;) I just think it'll _more code_ instead of terser code and _harder to understand_ than just parsing the messages. Also, we _count_ stuff, so in theory this is an infinite state machine, which might be a bit hard to model.

Reply:
https://codesandbox.io/s/hopeful-worker-9df2uo?file=/src/index.ts